kroxylicious-schema-validation/src/main/java/io/kroxylicious/proxy/filter/schema/ProduceValidationFilter.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.proxy.filter.schema;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.kafka.common.message.ProduceRequestData;
+import org.apache.kafka.common.message.ProduceResponseData;
+import org.apache.kafka.common.message.RequestHeaderData;
+import org.apache.kafka.common.message.ResponseHeaderData;
+import org.apache.kafka.common.protocol.Errors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.kroxylicious.proxy.filter.KrpcFilterContext;
+import io.kroxylicious.proxy.filter.ProduceRequestFilter;
+import io.kroxylicious.proxy.filter.ProduceResponseFilter;
+import io.kroxylicious.proxy.filter.schema.validation.ProduceRequestValidator;
+import io.kroxylicious.proxy.filter.schema.validation.RecordValidationFailure;
+import io.kroxylicious.proxy.filter.schema.validation.TopicValidationResult;
+import io.kroxylicious.proxy.filter.schema.validation.ValidationResult;
+
+import static java.util.stream.Collectors.joining;
+
+public class ProduceValidationFilter
+        implements ProduceRequestFilter, ProduceResponseFilter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ProduceValidationFilter.class);
+    private final ProduceRequestValidator validator;
+    private final Map<Integer, ValidationResult> correlatedResults = new HashMap<>();
+
+    public ProduceValidationFilter(ProduceRequestValidator validator) {
+        this.validator = validator;
+    }
+
+    @Override
+    public void onProduceRequest(RequestHeaderData header, ProduceRequestData request, KrpcFilterContext context) {
+        ValidationResult result = validator.validateRequest(request);
+        if (result.allTopicPartitionsInvalid()) {
+            LOGGER.warn("all topic-partitions for request contained invalid data: {}", result);
+            ProduceResponseData response = invalidateAllTopicPartitionProduceRequests(result);
+            context.forwardResponse(response);
+        }
+        else if (result.anyTopicPartitionInvalid()) {
```

Comment:
How would the `someInvalid` processing play with transactions?  Wouldn't we risk putting an application into an inconsistent state by sending only a subset of records?   I think at a bare minimum someInvalid processing would be an option a user would turn on.  I think reject the whole batch is safer default semantics.

Reply:
Great point. I think (but don't know for certain) that it should be fine. In the sense that the `INVALID_RECORD` error code is not considered retriable, so the producer should propagate it to the application, which should abort the transaction if it has proper error handling. 

We definitely want a test for this case, including that the `read_committed` consumer can't observe records published to other partitions in the same txn.

Reply:
I've changed it to invalidate the whole request if any topic-partition is invalid and there is a `transactionalId` present on the response, this way we save the broker the work of storing/forwarding the records only to have them invalidated later on abort.